lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -783,6 +783,162 @@ void LifetimeChecker::diagnoseInitError(const DIMemoryUse &Use,
+  if (!ctor)
+    return;
+
+  // If the initializer has restricted use of `self`, then no hops are needed.
+  if (isRestrictedActorInitSelf())
+    return;
+
+  // Even if there are no stored properties to initialize, we still need a hop.
+  if (TheMemory.getNumElements() == 0)
+    return injectHopAfterBuiltinInit(ctor,F.getEntryBlock());
+
+  // A mapping from initializing blocks to their in-availability set.
+  using Map = llvm::SmallDenseMap<SILBasicBlock*, AvailabilitySet, 8>;
+
+  /////
+  // Step 1: Find initializing blocks, which are blocks that contain a store
+  // to TheMemory that fully-initializes it, and build the Map.
+
+  Map initializingBlocks;
+  {
+    llvm::SmallDenseSet<SILBasicBlock *> seenBlocks; // processed blocks
+
+    // We determine whether a block is "initializing" by inspecting the "in" and
+    // "out" availability sets of the block. If the block goes from No / Partial
+    // "in" to Yes "out", then some instruction in the block caused TheMemory to
+    // become fully-initialized, so we record that block and its instructions
+    // that might be the ones that fully-initialized TheMemory.
+    for (auto const &entry : NonLoadUses) {
+      auto *inst = entry.getFirst();
+      auto *block = inst->getParent();
+
+      // TODO: we could probably skip use kinds that do not initialize memory.
+
+      // If we have already seen this block...
+      if (!seenBlocks.insert(block).second)
+        continue; // skip it!
+
+      // Determine if this block is initializing, that is:
+      //
+      //     merge(OutAvailability(block.pred1, block.pred2, ...)) != Yes
+      //               AND
+      //     OutAvailability(block)) = Yes
+
+      auto outSet = getBlockInfo(block).OutAvailability;
+      if (!outSet.isAllYes())
+        continue; // then this block never sees TheMemory fully-initialized.
+
+      AvailabilitySet inSet(outSet.size());
+      for (const auto &pred : block->getPredecessorBlocks())
+        inSet.mergeIn(getBlockInfo(pred).OutAvailability);
+
+      if (inSet.isAllYes())
+        continue; // then this block always receives TheMemory fully-initialized.
+
+      initializingBlocks.insert({block, inSet});
+    } // end loop
+
+#ifndef NDEBUG
+    // dump the data-structure we just built
+    for (auto &entry : initializingBlocks) {
+      auto &block = entry.getFirst();
+      llvm::dbgs() << "in " << block->getParent()->getName()
+                   << ", bb" << block->getDebugID()
+                   << " is an initializing block with in-availability:\t";
+      entry.getSecond().dump(llvm::dbgs());
+      llvm::dbgs() << "\n";
+    }
+#endif
+  }
+  ////// end of Step 1
+
+  // Step 2: Scan each block to find the first non-load use that
+  // fully-initializes TheMemory, using the Map from Step 1.
+
+  // Tracks status of each element of TheMemory as we scan through the block.
+  AvailabilitySet localAvail;
+
+  for (auto const& entry : initializingBlocks) {
```

Comment:
Iterating over a  pointer hash map is very dangerous because it can be easily introducing non-deterministic behavior. E.g, in this case, the order of the use list of the `hop_to_executor` operand would be non-deterministic.

Even if you make `initializingBlocks` an array, it would not solve the problem, because you are computing `initializingBlocks` by iterating over the `NonLoadUses` hash map.

You could just iterate over all basic blocks in the function to get a deterministic order. Complexity-wise this would be okay, because `injectActorHops()` is only done a single time for an actor constructor.

Reply:
Gotcha. I'll go over all the blocks and ask if they have non-load uses, and then scan each one, so that its more deterministic.